## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#95197f47-e53c-443f-9d02-729f245732b2) |
| mint-large-bucket.sh | :x: [more...](#ff30c9a2-1eb7-4898-a2d8-d88d7eab303e) |
| mint-dist-xl.sh | :x: [more...](#1d849e1c-20d2-4329-8713-aaf7299309d2) |
| mint-gateway-azure.sh | :x: [more...](#02114809-938b-4d5a-8487-9d4bd8a55660) |
| mint-zoned.sh | :x: [more...](#8cf67431-a2de-46ba-8566-529e5c1cfa03) |
<a name="02114809-938b-4d5a-8487-9d4bd8a55660"></a>
### 9963-b6eedd2/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c2.minio.io:30650
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 5c90da409c40:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 3 seconds
(3/15) Running aws-sdk-php tests ... done in 2 minutes and 39 seconds
(4/15) Running aws-sdk-ruby tests ... done in 17 seconds
(5/15) Running awscli tests ... done in 2 minutes and 47 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 3 minutes and 3 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 32 seconds
(9/15) Running minio-go tests ... done in 4 minutes and 52 seconds
(10/15) Running minio-java tests ... FAILED in 5 minutes and 17 seconds
{
  "name": "minio-java",
  "function": "removeObjects(RemoveObjectsArgs args)",
  "duration": 1269,
  "status": "FAIL",
  "error": "org.simpleframework.xml.core.ValueRequiredException: Unable to satisfy @org.simpleframework.xml.Element(name=\"Key\", type=void.class, data=false, required=true) on field 'name' private java.lang.String io.minio.messages.DeletedObject.name for class io.minio.messages.DeletedObject at line 2 >>> [io.minio.Xml.unmarshal(Xml.java:68), io.minio.MinioClient.deleteObjects(MinioClient.java:7354), io.minio.MinioClient$1$1.populate(MinioClient.java:2945), io.minio.MinioClient$1$1.hasNext(MinioClient.java:2977), FunctionalTest.removeObjects(FunctionalTest.java:1395), FunctionalTest.removeObjects_test1(FunctionalTest.java:1534), FunctionalTest.runTests(FunctionalTest.java:3812), FunctionalTest.main(FunctionalTest.java:4056)]"
}
(10/15) Running minio-js tests ... done in 2 minutes and 31 seconds
(11/15) Running minio-py tests ... done in 9 minutes and 55 seconds
(12/15) Running s3cmd tests ... done in 1 minutes and 21 seconds
(13/15) Running s3select tests ... done in 1 minutes and 5 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="8cf67431-a2de-46ba-8566-529e5c1cfa03"></a>
### 9963-b6eedd2/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev3.minio.io:31309
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp efabb7980666:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 16 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 48 seconds
(4/15) Running aws-sdk-ruby tests ... done in 14 seconds
(5/15) Running awscli tests ... done in 1 minutes and 56 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 1 minutes and 51 seconds
(8/15) Running minio-dotnet tests ... done in 2 minutes and 49 seconds
(9/15) Running minio-go tests ... FAILED in 2 minutes and 47 seconds
{
  "args": {
    "bucketName": "minio-go-test-k6ve05fhbdr1xxfe",
    "objectName": "k3n9vxyh9wwnfdfefg55xeylwssr0s",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 4954,
  "error": "The bucket you tried to delete is not empty. You must delete all versions in the bucket.",
  "function": "RemoveObjectsWithOptions(bucketName, objectsCh, opts)",
  "message": "Cleanup failed",
  "name": "minio-go: testRemoveObjectsWithOptions",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 5 minutes and 44 seconds
(10/15) Running minio-js tests ... done in 1 minutes and 0 seconds
(11/15) Running minio-py tests ... done in 5 minutes and 11 seconds
(12/15) Running s3cmd tests ... done in 23 seconds
(13/15) Running s3select tests ... done in 10 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="95197f47-e53c-443f-9d02-729f245732b2"></a>
### 9963-b6eedd2/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:30794
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 6ff33440b797:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... done in 1 minutes and 8 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... done in 23 seconds
(8/15) Running minio-dotnet tests ... done in 34 seconds
(9/15) Running minio-go tests ... FAILED in 57 seconds
{
  "args": {
    "bucketName": "minio-go-test-ilsx2fgxgsslpy49",
    "objectName": "5qbtq3l6aym45sd65ba4qnwod3lwna",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 4908,
  "error": "The bucket you tried to delete is not empty. You must delete all versions in the bucket.",
  "function": "RemoveObjectsWithOptions(bucketName, objectsCh, opts)",
  "message": "Cleanup failed",
  "name": "minio-go: testRemoveObjectsWithOptions",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 1 minutes and 16 seconds
(10/15) Running minio-js tests ... done in 47 seconds
(11/15) Running minio-py tests ... done in 3 minutes and 11 seconds
(12/15) Running s3cmd tests ... done in 19 seconds
(13/15) Running s3select tests ... done in 7 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="1d849e1c-20d2-4329-8713-aaf7299309d2"></a>
### 9963-b6eedd2/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:31634
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp f06006c59639:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... FAILED in 6 minutes and 10 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-ktocww1v163p6r",
    "expiry": 60000000000,
    "objectName1": "testObject1",
    "objectName2": "testObject2"
  },
  "duration": 368174,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 161E15B773D1E8EA, host id: ",
  "function": "testListObjects",
  "message": "AWS SDK Go listobjects expected to success but got XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 161E15B773D1E8EA, host id: ",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-java tests ... FAILED in 4 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-ktocww1v163p6r",
    "expiry": 60000000000,
    "objectName1": "testObject1",
    "objectName2": "testObject2"
  },
  "duration": 368174,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 161E15B773D1E8EA, host id: ",
  "function": "testListObjects",
  "message": "AWS SDK Go listobjects expected to success but got XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 161E15B773D1E8EA, host id: ",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-php tests ... done in 50 seconds
(2/15) Running aws-sdk-ruby tests ... FAILED in 3 seconds
{
  "name": "aws-sdk-ruby",
  "function": "listBuckets",
  "args": {},
  "duration": 2119.29,
  "error": "Server not initialized, please try again.",
  "status": "FAIL"
}
(2/15) Running awscli tests ... FAILED in 22 seconds
{
  "name": "awscli",
  "duration": 19997,
  "function": "delete_bucket\n",
  "status": "FAIL",
  "error": "fatal error: An error occurred (XMinioServerNotInitialized) when calling the ListObjectsV2 operation (reached max retries: 4): Server not initialized, please try again.remove_bucket failed: Unable to delete all objects in the bucket, bucket will not be deleted."
}
(2/15) Running healthcheck tests ... done in 0 seconds
(3/15) Running mc tests ... FAILED in 9 seconds
{
  "name": "mc",
  "duration": "588",
  "function": "test_get_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-22430 --quiet --no-color cp myminio/mc-test-bucket-10844/mc-test-object-16652 mc-test-object-16652.downloaded  >>> \nmc: <ERROR> Unable to validate source myminio/mc-test-bucket-10844/mc-test-object-16652"
}
(3/15) Running minio-dotnet tests ... FAILED in 33 seconds
{
  "name": "minio-dotnet : RemoveObject_Test1",
  "function": "Task RemoveObjectAsync(string bucketName, string objectName, CancellationToken cancellationToken = default(CancellationToken))",
  "args": {
    "bucketName": "miniodotnetnry9l7gyb48m495",
    "objectName": "bc*d%]@@$&"
  },
  "duration": 1159,
  "status": "FAIL",
  "alert": "",
  "message": "MinIO API responded with message=Server not initialized, please try again.",
  "error": "Minio.Exceptions.UnexpectedMinioException: MinIO API responded with message=Server not initialized, please try again.\n   at Minio.MinioClient.ParseErrorFromContent(IRestResponse response) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 538\n   at Minio.MinioClient.ParseError(IRestResponse response) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 431\n   at Minio.MinioClient.HandleIfErrorResponse(IRestResponse response, IEnumerable`1 handlers, DateTime startTime) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 605\n   at Minio.MinioClient.ExecuteTaskCoreAsync(IEnumerable`1 errorHandlers, IRestRequest request, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 405\n   at Minio.MinioClient.RemoveBucketAsync(String bucketName, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/BucketOperations.cs:line 139\n   at Minio.Functional.Tests.FunctionalTest.TearDown(MinioClient minio, String bucketName) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 381\n   at Minio.Functional.Tests.FunctionalTest.RemoveObject_Test1(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 2017"
}
(3/15) Running minio-go tests ... FAILED in 1 minutes and 49 seconds
{
  "args": {
    "bucketName": "minio-go-test-a2gzdiwpry9zfmrb",
    "region": "eu-west-1"
  },
  "duration": 108976,
  "error": "Server not initialized, please try again.",
  "function": "MakeBucket(bucketName, region)",
  "message": "MakeBucket failed",
  "name": "minio-go: testGetObjectClosedTwiceV2",
  "status": "FAIL"
}
(3/15) Running minio-java tests ... FAILED in 23 seconds
{
  "name": "minio-java",
  "function": "getPresignedObjectUrl(GetPresignedObjectUrlArgs args)",
  "duration": 146,
  "status": "FAIL",
  "error": "java.lang.Exception: failed to create object. Response: Response{protocol=http/1.1, code=503, message=Service Unavailable, url=http://minio-dev4.minio.io:31634/minio-java-test-fpjoft/minio-java-test-7sap1e?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20200703%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20200703T001728Z&X-Amz-Expires=604800&X-Amz-SignedHeaders=host&X-Amz-Signature=51a2a5e4bedc84d8354329e2a329feced4ed6439b12ef63711ad2938d9cffce7}, Response body: <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>XMinioServerNotInitialized</Code><Message>Server not initialized, please try again.</Message><Key>minio-java-test-7sap1e</Key><BucketName>minio-java-test-fpjoft</BucketName><Resource>/minio-java-test-fpjoft/minio-java-test-7sap1e</Resource><RequestId>161E15F24844F29A</RequestId><HostId>ff1a9896-fb29-42a6-a6d4-7c0ac66232be</HostId></Error> >>> [FunctionalTest.readObject(FunctionalTest.java:314), FunctionalTest.getPresignedObjectUrl_test1(FunctionalTest.java:1743), FunctionalTest.runTests(FunctionalTest.java:3802), FunctionalTest.main(FunctionalTest.java:4056)]"
}
(3/15) Running minio-js tests ... FAILED in 44 seconds
{
  "name": "minio-js",
  "function": "removeBucket(bucketName)",
  "args": "bucketName:minio-js-test-4551604b-784e-480a-b9d5-caae5cf55050-region",
  "duration": 12,
  "status": "FAIL",
  "error": "S3Error: Server not initialized, please try again. at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}
(3/15) Running minio-py tests ... FAILED in 14 minutes and 58 seconds
{
  "name": "minio-py:test_list_objects_v2",
  "function": "list_objects_v2(bucket_name, prefix, recursive, start_after)",
  "args": {
    "bucket_name": "minio-py-test-2a67404e-dd2c-499f-9f56-d16708ee0c71",
    "object_name": "966b3b52-c80f-48d1-9485-0201d240732e",
    "recursive": true
  },
  "duration": 604438,
  "message": "HTTPConnectionPool(host='minio-dev4.minio.io', port=31634): Max retries exceeded with url: /minio-py-test-2a67404e-dd2c-499f-9f56-d16708ee0c71/?list-type=2&prefix=&start-after= (Caused by ResponseError('too many 503 error responses',))",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1221, in test_list_objects_v2\n    for obj in objects:\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1113, in list_objects_v2\n    query=query)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 2000, in _url_open\n    preload_content=preload_content)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 336, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  [Previous line repeated 2 more times]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 805, in urlopen\n    retries = retries.increment(method, url, response=response, _pool=self)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 439, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minio-dev4.minio.io', port=31634): Max retries exceeded with url: /minio-py-test-2a67404e-dd2c-499f-9f56-d16708ee0c71/?list-type=2&prefix=&start-after= (Caused by ResponseError('too many 503 error responses',))\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2101, in main\n    test_list_objects_v2(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1228, in test_list_objects_v2\n    raise Exception(err)\nException: HTTPConnectionPool(host='minio-dev4.minio.io', port=31634): Max retries exceeded with url: /minio-py-test-2a67404e-dd2c-499f-9f56-d16708ee0c71/?list-type=2&prefix=&start-after= (Caused by ResponseError('too many 503 error responses',))\n",
  "status": "FAIL"
}
(3/15) Running s3cmd tests ... FAILED in 50 seconds
{
  "name": "s3cmd",
  "duration": "48014",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "WARNING: Retrying failed request: /s3cmd-test-object-23744 (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 3 sec...\nWARNING: Retrying failed request: /s3cmd-test-object-23744 (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 6 sec...\nWARNING: Retrying failed request: /s3cmd-test-object-23744 (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 9 sec...\nWARNING: Retrying failed request: /s3cmd-test-object-23744 (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 12 sec...\nWARNING: Retrying failed request: /s3cmd-test-object-23744 (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 15 sec...\nERROR: S3 error: 503 (XMinioServerNotInitialized): Server not initialized, please try again."
}
(3/15) Running s3select tests ... FAILED in 10 minutes and 5 seconds
{
  "name": "s3select:test_sql_operators_precedence",
  "function": "select_object_content(bucket_name, object_name, opts)",
  "args": {
    "total_tests": 14,
    "total_success": 10
  },
  "duration": 601063,
  "message": "HTTPConnectionPool(host='minio-dev4.minio.io', port=31634): Max retries exceeded with url: /s3select-test-d9e1762b-5b89-44d4-aa19-2fc7d50d5959/b0fbc693-0911-47b9-9da5-2bfc432e8b9e (Caused by ResponseError('too many 503 error responses',))",
  "error": "Traceback (most recent call last):\n  File \"./tests.py\", line 54, in main\n    test_sql_operators_precedence(client, log_output)\n  File \"/mint/run/core/s3select/sql_ops.py\", line 163, in test_sql_operators_precedence\n    raise select_err\n  File \"/mint/run/core/s3select/sql_ops.py\", line 161, in test_sql_operators_precedence\n    test_sql_expressions(client, json_testfile, tests, log_output)\n  File \"/mint/run/core/s3select/sql_ops.py\", line 89, in test_sql_expressions\n    input_serialization, output_serialization, tests, log_output)\n  File \"/mint/run/core/s3select/sql_ops.py\", line 74, in test_sql_expressions_custom_input_output\n    client.remove_object(bucket_name, object_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1171, in remove_object\n    object_name=object_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 2000, in _url_open\n    preload_content=preload_content)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 336, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  [Previous line repeated 2 more times]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 805, in urlopen\n    retries = retries.increment(method, url, response=response, _pool=self)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 439, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minio-dev4.minio.io', port=31634): Max retries exceeded with url: /s3select-test-d9e1762b-5b89-44d4-aa19-2fc7d50d5959/b0fbc693-0911-47b9-9da5-2bfc432e8b9e (Caused by ResponseError('too many 503 error responses',))\n",
  "status": "FAIL"
}
(3/15) Running security tests ... done in 0 seconds

Executed 3 out of 15 tests successfully.
```
<a name="ff30c9a2-1eb7-4898-a2d8-d88d7eab303e"></a>
### 9963-b6eedd2/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:32613
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 34e123c62216:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 12 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 47 seconds
(4/15) Running aws-sdk-ruby tests ... done in 12 seconds
(5/15) Running awscli tests ... done in 2 minutes and 44 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... FAILED in 45 seconds
{
  "name": "mc",
  "duration": "11662",
  "function": "test_get_object_multipart",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-5254 --quiet --no-color cp myminio/mc-test-bucket-15296/mc-test-object-14290 mc-test-object-14290.downloaded  >>> \n`myminio/mc-test-bucket-15296/mc-test-object-14290` -> `mc-test-object-14290.downloaded`\nmc: <ERROR> Failed to copy `http://minio-dev7.minio.io:32613/mc-test-bucket-15296/mc-test-object-14290`. Input reader closed pre-maturely. Expected `68157440` bytes, but only received `52428800` bytes.\nTotal: 0 B, Transferred: 50.00 MiB, Speed: 5.91 MiB/s"
}
(7/15) Running minio-dotnet tests ... done in 2 minutes and 36 seconds
(8/15) Running minio-go tests ... FAILED in 2 minutes and 22 seconds
{
  "args": {
    "bucketName": "minio-go-test-2sxl0dz02433m2mr",
    "objectName": "cezuu6noksu14r04oaizspl6aybttj",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 3785,
  "error": "The bucket you tried to delete is not empty. You must delete all versions in the bucket.",
  "function": "RemoveObjectsWithOptions(bucketName, objectsCh, opts)",
  "message": "Cleanup failed",
  "name": "minio-go: testRemoveObjectsWithOptions",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... done in 3 minutes and 44 seconds
(9/15) Running minio-js tests ... done in 1 minutes and 2 seconds
(10/15) Running minio-py tests ... done in 4 minutes and 52 seconds
(11/15) Running s3cmd tests ... done in 23 seconds
(12/15) Running s3select tests ... done in 14 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/http/dial_linux.go
+++ b/cmd/http/dial_linux.go
@@ -23,29 +23,51 @@ import (
 	"net"
 	"syscall"
 	"time"
+
+	"golang.org/x/sys/unix"
 )
 
-const (
-	// TCPFastOpenConnect sets the underlying socket to use
-	// the TCP fast open connect. This feature is supported
-	// since Linux 4.11.
-	TCPFastOpenConnect = 30
-)
+func setTCPParameters(c syscall.RawConn) error {
+	return c.Control(func(fdPtr uintptr) {
+		// got socket file descriptor to set parameters.
+		fd := int(fdPtr)
+
+		// Enable TCP fast connect
+		// TCPFastOpenConnect sets the underlying socket to use
+		// the TCP fast open connect. This feature is supported
+		// since Linux 4.11.
+		_ = syscall.SetsockoptInt(int(fd), syscall.IPPROTO_TCP, unix.TCP_FASTOPEN_CONNECT, 1)
+
+		// The time (in seconds) the connection needs to remain idle before
+		// TCP starts sending keepalive probes, set this to 5 secs
+		// system defaults to 7200 secs!!!
+		_ = syscall.SetsockoptInt(int(fd), syscall.IPPROTO_TCP, syscall.TCP_KEEPIDLE, 5)
+
+		// Number of probes.
+		// ~ cat /proc/sys/net/ipv4/tcp_keepalive_probes (defaults to 9, we reduce it to 5)
+		// 9
+		_ = syscall.SetsockoptInt(int(fd), syscall.IPPROTO_TCP, syscall.TCP_KEEPCNT, 5)
+
+		// Wait time after successful probe in seconds.
+		// ~ cat /proc/sys/net/ipv4/tcp_keepalive_intvl (defaults to 75 secs, we reduce it to 2 secs)
+		// 75
+		_ = syscall.SetsockoptInt(int(fd), syscall.IPPROTO_TCP, syscall.TCP_KEEPINTVL, 2)
+
+		// Set TCP_USER_TIMEOUT to TCP_KEEPIDLE + TCP_KEEPINTVL * TCP_KEEPCNT.
+		_ = syscall.SetsockoptInt(int(fd), syscall.IPPROTO_TCP, unix.TCP_USER_TIMEOUT, 15)
+	})
+}
 
 // DialContext is a function to make custom Dial for internode communications
 type DialContext func(ctx context.Context, network, address string) (net.Conn, error)
 
 // NewCustomDialContext setups a custom dialer for internode communications
-func NewCustomDialContext(dialTimeout, dialKeepAlive time.Duration) DialContext {
+func NewCustomDialContext(dialTimeout time.Duration) DialContext {
 	return func(ctx context.Context, network, addr string) (net.Conn, error) {
 		dialer := &net.Dialer{
-			Timeout:   dialTimeout,
-			KeepAlive: dialKeepAlive,
+			Timeout: dialTimeout,
 			Control: func(network, address string, c syscall.RawConn) error {
-				return c.Control(func(fd uintptr) {
-					// Enable TCP fast connect
-					_ = syscall.SetsockoptInt(int(fd), syscall.IPPROTO_TCP, TCPFastOpenConnect, 1)
-				})
+				return setTCPParameters(c)
 			},
 		}
 		return dialer.DialContext(ctx, network, addr)

--- a/cmd/http/dial_others.go
+++ b/cmd/http/dial_others.go
@@ -21,18 +21,23 @@ package http
 import (
 	"context"
 	"net"
+	"syscall"
 	"time"
 )
+
+// TODO: if possible implement for non-linux platforms, not a priority at the moment
+func setTCPParameters(c syscall.RawConn) error {
+	return nil
+}
 
 // DialContext is a function to make custom Dial for internode communications
 type DialContext func(ctx context.Context, network, address string) (net.Conn, error)
 
 // NewCustomDialContext configures a custom dialer for internode communications
-func NewCustomDialContext(dialTimeout, dialKeepAlive time.Duration) DialContext {
+func NewCustomDialContext(dialTimeout time.Duration) DialContext {
 	return func(ctx context.Context, network, addr string) (net.Conn, error) {
 		dialer := &net.Dialer{
-			Timeout:   dialTimeout,
-			KeepAlive: dialKeepAlive,
+			Timeout: dialTimeout,
 		}
 		return dialer.DialContext(ctx, network, addr)
 	}

--- a/cmd/http/listener.go
+++ b/cmd/http/listener.go
@@ -23,7 +23,6 @@ import (
 	"os"
 	"sync"
 	"syscall"
-	"time"
 )
 
 type acceptResult struct {
@@ -33,11 +32,10 @@ type acceptResult struct {
 
 // httpListener - HTTP listener capable of handling multiple server addresses.
 type httpListener struct {
-	mutex               sync.Mutex         // to guard Close() method.
-	tcpListeners        []*net.TCPListener // underlaying TCP listeners.
-	acceptCh            chan acceptResult  // channel where all TCP listeners write accepted connection.
-	doneCh              chan struct{}      // done channel for TCP listener goroutines.
-	tcpKeepAliveTimeout time.Duration
+	mutex        sync.Mutex         // to guard Close() method.
+	tcpListeners []*net.TCPListener // underlaying TCP listeners.
+	acceptCh     chan acceptResult  // channel where all TCP listeners write accepted connection.
+	doneCh       chan struct{}      // done channel for TCP listener goroutines.
 }
 
 // isRoutineNetErr returns true if error is due to a network timeout,
@@ -83,10 +81,10 @@ func (listener *httpListener) start() {
 
 	// Closure to handle single connection.
 	handleConn := func(tcpConn *net.TCPConn, doneCh <-chan struct{}) {
-		// Tune accepted TCP connection.
-		tcpConn.SetKeepAlive(true)
-		tcpConn.SetKeepAlivePeriod(listener.tcpKeepAliveTimeout)
-
+		rawConn, err := tcpConn.SyscallConn()
+		if err == nil {
+			setTCPParameters(rawConn)
+		}
 		send(acceptResult{tcpConn, nil}, doneCh)
 	}
 
@@ -167,8 +165,7 @@ func (listener *httpListener) Addrs() (addrs []net.Addr) {
 // httpListener is capable to
 // * listen to multiple addresses
 // * controls incoming connections only doing HTTP protocol
-func newHTTPListener(serverAddrs []string,
-	tcpKeepAliveTimeout time.Duration) (listener *httpListener, err error) {
+func newHTTPListener(serverAddrs []string) (listener *httpListener, err error) {
 
 	var tcpListeners []*net.TCPListener
 
@@ -201,8 +198,7 @@ func newHTTPListener(serverAddrs []string,
 	}
 
 	listener = &httpListener{
-		tcpListeners:        tcpListeners,
-		tcpKeepAliveTimeout: tcpKeepAliveTimeout,
+		tcpListeners: tcpListeners,
 	}
 	listener.start()
 

--- a/cmd/http/listener_test.go
+++ b/cmd/http/listener_test.go
@@ -151,7 +151,6 @@ func TestNewHTTPListener(t *testing.T) {
 	for _, testCase := range testCases {
 		listener, err := newHTTPListener(
 			testCase.serverAddrs,
-			testCase.tcpKeepAliveTimeout,
 		)
 
 		if !testCase.expectedErr {
@@ -185,7 +184,6 @@ func TestHTTPListenerStartClose(t *testing.T) {
 	for i, testCase := range testCases {
 		listener, err := newHTTPListener(
 			testCase.serverAddrs,
-			time.Duration(0),
 		)
 		if err != nil {
 			t.Fatalf("Test %d: error: expected = <nil>, got = %v", i+1, err)
@@ -225,7 +223,6 @@ func TestHTTPListenerAddr(t *testing.T) {
 	for i, testCase := range testCases {
 		listener, err := newHTTPListener(
 			testCase.serverAddrs,
-			time.Duration(0),
 		)
 		if err != nil {
 			t.Fatalf("Test %d: error: expected = <nil>, got = %v", i+1, err)
@@ -262,7 +259,6 @@ func TestHTTPListenerAddrs(t *testing.T) {
 	for i, testCase := range testCases {
 		listener, err := newHTTPListener(
 			testCase.serverAddrs,
-			time.Duration(0),
 		)
 		if err != nil {
 			t.Fatalf("Test %d: error: expected = <nil>, got = %v", i+1, err)

--- a/cmd/http/server.go
+++ b/cmd/http/server.go
@@ -38,9 +38,6 @@ const (
 	// DefaultShutdownTimeout - default shutdown timeout used for graceful http server shutdown.
 	DefaultShutdownTimeout = 5 * time.Second
 
-	// DefaultTCPKeepAliveTimeout - default TCP keep alive timeout for accepted connection.
-	DefaultTCPKeepAliveTimeout = 30 * time.Second
-
 	// DefaultMaxHeaderBytes - default maximum HTTP header size in bytes.
 	DefaultMaxHeaderBytes = 1 * humanize.MiByte
 )
@@ -48,13 +45,12 @@ const (
 // Server - extended http.Server supports multiple addresses to serve and enhanced connection handling.
 type Server struct {
 	http.Server
-	Addrs               []string      // addresses on which the server listens for new connection.
-	ShutdownTimeout     time.Duration // timeout used for graceful server shutdown.
-	TCPKeepAliveTimeout time.Duration // timeout used for underneath TCP connection.
-	listenerMutex       sync.Mutex    // to guard 'listener' field.
-	listener            *httpListener // HTTP listener for all 'Addrs' field.
-	inShutdown          uint32        // indicates whether the server is in shutdown or not
-	requestCount        int32         // counter holds no. of request in progress.
+	Addrs           []string      // addresses on which the server listens for new connection.
+	ShutdownTimeout time.Duration // timeout used for graceful server shutdown.
+	listenerMutex   sync.Mutex    // to guard 'listener' field.
+	listener        *httpListener // HTTP listener for all 'Addrs' field.
+	inShutdown      uint32        // indicates whether the server is in shutdown or not
+	requestCount    int32         // counter holds no. of request in progress.
 }
 
 // GetRequestCount - returns number of request in progress.
@@ -72,13 +68,11 @@ func (srv *Server) Start() (err error) {
 	handler := srv.Handler // if srv.Handler holds non-synced state -> possible data race
 
 	addrs := set.CreateStringSet(srv.Addrs...).ToSlice() // copy and remove duplicates
-	tcpKeepAliveTimeout := srv.TCPKeepAliveTimeout
 
 	// Create new HTTP listener.
 	var listener *httpListener
 	listener, err = newHTTPListener(
 		addrs,
-		tcpKeepAliveTimeout,
 	)
 	if err != nil {
 		return err
@@ -204,9 +198,8 @@ func NewServer(addrs []string, handler http.Handler, getCert certs.GetCertificat
 	}
 
 	httpServer := &Server{
-		Addrs:               addrs,
-		ShutdownTimeout:     DefaultShutdownTimeout,
-		TCPKeepAliveTimeout: DefaultTCPKeepAliveTimeout,
+		Addrs:           addrs,
+		ShutdownTimeout: DefaultShutdownTimeout,
 	}
 	httpServer.Handler = handler
 	httpServer.TLSConfig = tlsConfig

--- a/cmd/http/server_test.go
+++ b/cmd/http/server_test.go
@@ -73,10 +73,6 @@ func TestNewServer(t *testing.T) {
 			t.Fatalf("Case %v: server.ShutdownTimeout: expected: %v, got: %v", (i + 1), DefaultShutdownTimeout, server.ShutdownTimeout)
 		}
 
-		if server.TCPKeepAliveTimeout != DefaultTCPKeepAliveTimeout {
-			t.Fatalf("Case %v: server.TCPKeepAliveTimeout: expected: %v, got: %v", (i + 1), DefaultTCPKeepAliveTimeout, server.TCPKeepAliveTimeout)
-		}
-
 		if server.MaxHeaderBytes != DefaultMaxHeaderBytes {
 			t.Fatalf("Case %v: server.MaxHeaderBytes: expected: %v, got: %v", (i + 1), DefaultMaxHeaderBytes, server.MaxHeaderBytes)
 		}

--- a/cmd/update.go
+++ b/cmd/update.go
@@ -409,7 +409,7 @@ const updateTimeout = 10 * time.Second
 func getUpdateTransport(timeout time.Duration) http.RoundTripper {
 	var updateTransport http.RoundTripper = &http.Transport{
 		Proxy:                 http.ProxyFromEnvironment,
-		DialContext:           xhttp.NewCustomDialContext(timeout, timeout),
+		DialContext:           xhttp.NewCustomDialContext(timeout),
 		IdleConnTimeout:       timeout,
 		TLSHandshakeTimeout:   timeout,
 		ExpectContinueTimeout: timeout,

--- a/cmd/utils.go
+++ b/cmd/utils.go
@@ -454,7 +454,7 @@ func newCustomHTTPTransport(tlsConfig *tls.Config, dialTimeout time.Duration) fu
 	// https://golang.org/pkg/net/http/#Transport documentation
 	tr := &http.Transport{
 		Proxy:                 http.ProxyFromEnvironment,
-		DialContext:           xhttp.NewCustomDialContext(dialTimeout, 15*time.Second),
+		DialContext:           xhttp.NewCustomDialContext(dialTimeout),
 		MaxIdleConnsPerHost:   16,
 		MaxIdleConns:          16,
 		IdleConnTimeout:       1 * time.Minute,

